ada_feeding/ada_feeding/decorators/on_preempt.py

```diff
@@ -0,0 +1,109 @@
+"""
+NOTE: This is a multi-tick version of the decorator discussed in
+https://github.com/splintered-reality/py_trees/pull/427 . Once a
+multi-tick version of that decorator is merged into py_trees, this
+decorator should be removed in favor of the main py_trees one.
+"""
+
+import time
+import typing
+
+from py_trees import behaviour, common
+from py_trees.decorators import Decorator
+
+
+class OnPreempt(Decorator):
```

Comment:
To clarify, when a tree is preempted, `Behavior.stop(INVALID)` is called on *every* node in the tree (or, more accurately, it is called on the parent node, which for all Decorators/Composites calls stop(INVALID) on all children that are not already INVALID).

There's some potentially unintended behavior here.

(1) You address this below, but of course, this node needs to be ticked at least once to "activate". IT is best to document that here.

(2) Once this node is activated (i.e. will catch preemptions), it appears to active *forever*, causing the preemption functionality to bleed into the entire tree. This wouldn't be caught in a unit test that only considers "local" preemptions.


Here is how I would fix this: Make this a true decorator (with the preemption subtree not a true "child") and have the preemption protection active only on the subtree.

Stopping the review here as this affects the rest of the code.

Reply:
Hmm, I don't understand your proposed fix, let's discuss more in lab today.